### ISSUE TITLE
Logging

### ISSUE THREAD

**Issue by Lukasa** [python-hyper/hyper-h2#11]
It's really important that this state machine be able to log, but it's also really important that we don't break our rule about I/O. This means we can't just transparently use Python's standard `logging` module: on systems like gevent this will implicitly cause context switches, and on systems like Twisted everything goes very wrong.

The best approach is probably to allow injection of a logger that conforms to the rules of this module (must not do I/O, must not context switch), and push that problem outwards. This logger needs to use the same interface as the standard library's `logging` module, but otherwise everything else is fine.

This is tricky, but we can also defer it from 1.0.0 I think.

**Comment by fredthomsen** [comment]
I am assume the context switch you are referring to here would occur if the `logging` module was using `syslog` and thus a network call was made.

**Comment by Lukasa** [comment]
That's the most likely scenario, yeah, but you could construct context-switching file-IO if you wanted to as well.

**Comment by fredthomsen** [comment]
Maybe I am being a bit dense here, but how are you logging without I/O?  Are you implying some sort of deferred/buffered I/O that would only write to memory during execution that dump to file upon program termination/some event/an exception?

**Comment by Lukasa** [comment]
Essentially the user would need to provide a buffer for hyper-h2 to log to, and would then be responsible for emitting that buffer as appropriate for its I/O model. Same as the standard logic hyper-h2 uses for the actual I/O stuff really.

**Comment by fredthomsen** [comment]
Obviously I am interested in taking a stab at this.  I was thinking it would consist of:

- A sample logger that conforms to these requirements.
- A way to inject said logger
- And obviously usage of the logger

**Comment by Lukasa** [comment]
I think we'd go a different way, actually.

We'd want to provide *examples* of the logger in the examples directory, but I don't think we'd want to provide one in our product code: it's not really necessary and we'd need multiples. However, we would need a default logger that is a no-op so that the product code isn't dirtied up with conditional checks.

As to the other two cases, I was envisioning a submodule (`h2.logging`) which has the following methods:

- `h2.logging.inject_logger()`
- `h2.logging.extract_logger()`
- `h2.logging.log()`

Then the rest of the code can simply do `import h2.logging as log; log.log()` as needed.

How does that sound?

**Comment by rbtcollins** [comment]
So a few thoughts here.

Global state is icky. Globally injecting a logger means that all my instances of h2 must agree that that is how logging should be done; this loses one of the major benefits of being decoupled from the world - we're no longer an isolated state machine. So if we're going to do something custom lets pass it in the top rather than tying all the backends together with invisible state.

Conditional evaluation can end up being expensive too; if there's lots of logging happening we may still want conditionals in h2 - but fine to defer that until needed.

My suggestion: just log with the system standard logging module. Someone using gevent can use appropriate handlers to avoid context switching  - and discard/fall back to console/stderr for switches that would try to reenter the h2 state machine that invoked the log in the first place. - handling reentrancy in logging is a known thing, for all that its ugly.
This would give: a) works out of the box and b) there's a clean way to make it work in any given environment. It does do the same backend state tying I whine about earlier, but at zero dev cost :P.

**Comment by Lukasa** [comment]
@rbtcollins I'm unconvinced by the idea that someone using gevent will do the right thing here, mostly because of my own personal experience having *actually done it wrong*. The added downside of using the standard library logging framework is that if we enable the logging by default then any library built on hyper-h2 that does not expect context switches inside a hyper-h2 call may still get them accidentally, due to a misunderstanding between application developer and library author (the library author considers hyper-h2 an implementation detail, and the application developer wants to log via syslog with logging being a context switch event). This breaks *all the things* in a pretty profound way.

**Comment by rbtcollins** [comment]
Ok; that fits into the same category of failure modes as two distinct libraries using h2 in one app, and global state for logging: so lets have the API be that logging is provided to h2, and any library using h2 needs to provide a logger compatible with its needs.

**Comment by Lukasa** [comment]
@rbtcollins What would that look like exactly?

**Comment by rbtcollins** [comment]
```
# Use system logger
conn = H2Connection(client_side=False, logger=logging)
```

**Comment by Lukasa** [comment]
Yup, that can work. We'll need to use the new config object, but that's for the best, because it's passed to most of the relevant hyper-h2 sub-objects anyway.

**Comment by fredthomsen** [comment]
The `H2Configuration` would get passed the relevant logger then and the connections can use it appropriately.  I gather any logger should still use the same API as the standard library logger as you indicated in the first post; however, now that we are injecting a logger into each connection I would think you would rather have a default implementation rather than an example.

**Comment by Lukasa** [comment]
@fredthomsen At this point we've moved to expecting a standard-library logger directly, so API compatibility is a non-issue.

In point of fact, we'd probably default to the standard library's NullLogger, which neatly resolves our conditionals problem.

**Comment by fredthomsen** [comment]
Sounds good. I'll move ahead with this.

**Comment by Lukasa** [comment]
Resolved by #436.
